From: Andrew Morton <akpm@linux-foundation.org>
Subject: linux-next-git-rejects

Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 drivers/platform/x86/intel/Kconfig |    3 ---
 1 file changed, 3 deletions(-)

--- a/drivers/platform/x86/intel/Kconfig~linux-next-git-rejects
+++ a/drivers/platform/x86/intel/Kconfig
@@ -170,8 +170,6 @@ config INTEL_UNCORE_FREQ_CONTROL
 
 	  To compile this driver as a module, choose M here: the module
 	  will be called intel-uncore-frequency.
-<<<<<<< HEAD
-=======
 
 config INTEL_VSEC
 	tristate "Intel Vendor Specific Extended Capabilities Driver"
@@ -183,4 +181,3 @@ config INTEL_VSEC
 
 	  To compile this driver as a module, choose M here: the module will
 	  be called intel_vsec.
->>>>>>> linux-next/akpm-base
_
