From: Andrew Morton <akpm@linux-foundation.org>
Subject: linux-next-git-rejects

Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 sound/soc/codecs/cs42l42.c |    3 ---
 1 file changed, 3 deletions(-)

--- a/sound/soc/codecs/cs42l42.c~linux-next-git-rejects
+++ a/sound/soc/codecs/cs42l42.c
@@ -844,8 +844,6 @@ static int cs42l42_pcm_hw_params(struct
 	if (channels == 1)
 		cs42l42->bclk *= 2;
 
-<<<<<<< HEAD
-=======
 	/*
 	 * Assume 24-bit samples are in 32-bit slots, to prevent SCLK being
 	 * more than assumed (which would result in overclocking).
@@ -853,7 +851,6 @@ static int cs42l42_pcm_hw_params(struct
 	if (params_width(params) == 24)
 		cs42l42->bclk = (cs42l42->bclk / 3) * 4;
 
->>>>>>> linux-next/akpm-base
 	switch(substream->stream) {
 	case SNDRV_PCM_STREAM_CAPTURE:
 		if (channels == 2) {
_
