From: Andrew Morton <akpm@linux-foundation.org>
Subject: linux-next-git-rejects

Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 kernel/printk/sysctl.c |   12 ------------
 1 file changed, 12 deletions(-)

--- a/kernel/printk/sysctl.c~linux-next-git-rejects
+++ a/kernel/printk/sysctl.c
@@ -11,18 +11,6 @@
 
 static const int ten_thousand = 10000;
 
-<<<<<<< HEAD
-static int proc_dointvec_minmax_sysadmin(struct ctl_table *table, int write,
-				void *buffer, size_t *lenp, loff_t *ppos)
-{
-	if (write && !capable(CAP_SYS_ADMIN))
-		return -EPERM;
-
-	return proc_dointvec_minmax(table, write, buffer, lenp, ppos);
-}
-
-=======
->>>>>>> linux-next/akpm-base
 static struct ctl_table printk_sysctls[] = {
 	{
 		.procname	= "printk",
_
